`_score` should not be a reserved attribute in ES|QL

tteofili opened the pull request:
Drop checks in ES|QL `Verifier` that ensured `_score` is a reserved attribute.
These checks are too restrictive, in addition to that relaxing them would make handling of `_score` consistent with other (metadata) attributes and fields.

fixes #118460

tteofili committed:
_score not reserved attribute

tteofili committed:
Update docs/changelog/118435.yaml

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

tteofili committed:
reverted unrelated changes

tteofili committed:
Merge branch 'esql_score_reserved' of github.com:tteofili/elasticsearch into esql_score_reserved

tteofili committed:
Merge branch 'main' into esql_score_reserved

ioanatia reviewed (commented):
The changes look right with me!
Can we get a CSV test just to prevent any future regressions with allowing `_score` to be overwritten?
Something like what you had in the test before: `"from foo metadata _score | where qstr(\"bar\") | eval _score = _score + 1"`
and another one where `_score` receives a different type:

`from foo metadata _score | where qstr(\"bar\") | eval _score = "foobar"`

with that, I think this should be good to go!

ioanatia commented:
also what would happen if `_score` is an actual indexed field? have we tested that?
we don't have to fix it here, but can be a follow up in the meta issue https://github.com/elastic/elasticsearch/issues/116599

tteofili committed:
add scoring manipulation/override csv tests

tteofili committed:
Merge branch 'esql_score_reserved' of github.com:tteofili/elasticsearch into esql_score_reserved

tteofili committed:
Merge branch 'main' of github.com:elastic/elasticsearch into esql_score_reserved

tteofili commented:
> also what would happen if `_score` is an actual indexed field? have we tested that?

added an item in #116599 

> an we get a CSV test just to prevent any future regressions with allowing _score to be overwritten?

added a couple of tests for score manipulation and override in [0ff6fb3](https://github.com/elastic/elasticsearch/pull/118435/commits/0ff6fb39a6a24236f9592c9125046fc96340f33e)

ioanatia commented:
> > an we get a CSV test just to prevent any future regressions with allowing _score to be overwritten?

> added a couple of tests for score manipulation and override in [0ff6fb3](https://github.com/elastic/elasticsearch/pull/118435/commits/0ff6fb39a6a24236f9592c9125046fc96340f33e)

I did not realize this would actually create a problem for bwc - feel free to remove them and then we can add them later after the changes make it in 8.18.
The alternatives are:

- mute the tests you just added
- add a new `EsqlCapability` to signal that `_score` is no longer reserved - this is too heavy, let's not do it

tteofili committed:
mute score manipulation csv tests for bwc tests (temporarily)

tteofili committed:
Merge branch 'main' of github.com:elastic/elasticsearch into esql_score_reserved

elasticsearchmachine commented:
Hi @tteofili, I've updated the changelog YAML for you.

tteofili committed:
Update docs/changelog/118435.yaml

tteofili committed:
Merge branch 'main' into esql_score_reserved

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 118435`

ioanatia commented on muted-tests.yml:
```diff
@@ -180,6 +180,12 @@ tests:
 - class: "org.elasticsearch.xpack.esql.qa.mixed.MixedClusterEsqlSpecIT"
   method: "test {scoring.*}"
   issue: https://github.com/elastic/elasticsearch/issues/117641
+- class: "org.elasticsearch.xpack.esql.qa.mixed.MultilusterEsqlSpecIT"
```
there is a typo here - also looks like a duplicate with the previous muted test (line 180)

tteofili replied:
oh no! dummy PR to fix that https://github.com/elastic/elasticsearch/pull/118665

tteofili commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/118568">](https://github.com/elastic/elasticsearch/pull/118568)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.3 -->
